src/main/kotlin/au/com/timmutton/redexplugin/RedexDownloadTask.kt

```diff
@@ -0,0 +1,114 @@
+package au.com.timmutton.redexplugin
+
+import java.util.regex.Pattern
+import java.util.regex.Matcher
+import java.nio.file.Files
+import java.nio.charset.Charset
+import java.io.File
+import java.io.FileOutputStream
+import java.io.FileNotFoundException
+import java.io.IOException
+import java.io.ByteArrayOutputStream
+import com.github.kittinunf.fuel.Fuel
+import com.github.kittinunf.result.Result
+import org.apache.tools.ant.taskdefs.condition.Os
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.TaskAction
+
+
+open class RedexDownloadTask : DefaultTask() {
+    var requestedVersion : String? = "latest"
+    public val buildDir = project.buildDir.getPath()
+
+    // Returns the location where redex will be downloaded
+    // after this task is run
+    public fun initialise(ext : RedexExtension) : File? {
+        requestedVersion = ext.version
+        val (redex, _) = getRedexExecutableFile()
+        return redex
+    }
+
+    @TaskAction
+    // download redex from the github release page
+    public fun run() {
+        if (requestedVersion == null) {
+            return
+        }
+        val (redex, url) = getRedexExecutableFile()
+        if (redex == null || url == null) {
+            return
+        }
+        if (!redex.exists()) {
+            // Only download if we don't already have it
+            download(url, redex)
+            redex.setExecutable(true)
+        }
+    }
+
+    // if dest is null, return the data
+    // if dest is not null, write it to dest
+    private fun download(url : String, dest : File?) : ByteArray? {
+        var output : ByteArray? = null
+        val (_,_, result) = Fuel.get(url).response()
+        val (data : ByteArray?, error) = result
+        if (error != null) {
+            throw error
+        }
+        if (data == null) {
+            throw IOException("No data from download")
+        }
+        if (dest == null) {
+            output = data
+        } else {
+            dest.writeBytes(data)
+        }
+        return output
```

Comment:
If dest is non-null, this wont be returning the data you downloaded

Reply:
That was intentional, but yeah this is confusing and error prone. I'll change it